### ISSUE TITLE
Releasing the stable Amidst v4.0

### ISSUE THREAD

**Issue by stefandollase** [toolbox4minecraft/amidst#64]
I just created another pre-release v4.0-beta8. It is the release candidate for the stable v4.0. This means, that beta8 is feature complete and that only bugs will be fixed until I create the stable release.

I would like to create the stable release quite soon, however we have to wait for the stable Minecraft 1.9 release for this. The reason for this is, that we need to extract the so called magic string from the Minecraft jar file, in order for Amidst to recognise it properly. So, if Amidst v4.0 should recognise Minecraft 1.9, we have to defer the release until 1.9 is released. However, this will probably be quite soon, so this should not be that big of a deal.

If you think there is something that really has to go into v4.0, let me know right here. If it is a bug-report, just create a new issue for it.

---

Tasks to do before releasing Amidst v4.0:

- [x] update cached version list: @stefandollase
- [x] update enum `RecognisedVersion`: @stefandollase

Tasks to do to release Amidst v4.0:

- [x] follow the instructions in [Creating a Release](https://github.com/toolbox4minecraft/amidst/wiki/Creating-a-Release): @stefandollase
- [x] update the update document: @stefandollase

Tasks to do after Amidst v4.0 was released:

- [x] update the screenshots
- [x] create [new minecraftforum thread](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-tools/2626547-amidst-has-found-a-new-home) in [Minecraft Tools](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-tools): @stefandollase
- [x] insert link to minecraftforum thread into the README: @stefandollase
- [x] update the README of this repository: @stefandollase
- [x] create [new minecraftforum thread](http://www.minecraftforum.net/forums/minecraft-discussion/seeds/2627219-amidst-now-supports-minecraft-1-9-and-has-a-new) in [Seeds](http://www.minecraftforum.net/forums/minecraft-discussion/seeds): @stefandollase
- [x] update information in the [minecraft wiki](http://minecraft.gamepedia.com/Programs_and_editors/Mapping): @stefandollase
- [x] ask @skiphs to complete the [tasks marked with a checkbox](https://github.com/skiphs/AMIDST/issues/92#issuecomment-183963569): @stefandollase
- [x]  create [thread](https://www.reddit.com/r/Minecraft/comments/48e3he/amidst_for_19_now_available_locate_igloos/) in [/r/minecraft](https://www.reddit.com/r/minecraft), if it does not appear by itself: @Treer

Tasks to do after Amidst v4.1 was released:

* change update document of AmidstExporter: @Treer 
* refer to this repository from the README of AmidstExporter: @Treer

**Comment by Treer** [comment]
:+1: 

Waiting for that magic string is the only thing I'm aware of remaining for a v4.0 stable.

**Comment by stefandollase** [comment]
I guess we have a new release candidate: Amidst v4.0-beta10

#40 will be a known issue and #43 should be fixed.

Also, I wrote up a list of things to do for the release. Can you check whether I missed something?

**Comment by Treer** [comment]
A couple of days after minecraftforum I suggest a post to reddit/r/minecraft as well. It's not as important as minecraftforum because in a few days the post will be forgotten but many people there don't read minecraftforum.

It will either drive traffic in a big short hit for a day or two, or sink like a stone :)

It also might get posted there without our help.

**Comment by stefandollase** [comment]
These are good points. I added them to the original post.

I guess we do not need a toolbox4minecraft reddit account, do we? The reddit post is probably short-lived. Since I do not have a reddit account: Can you create the post @Treer ?

**Comment by BigAlanM** [comment]
FYI - Minecraft 1.9 is on the street.

**Comment by stefandollase** [comment]
Thanks, I am aware of that.

**Comment by Treer** [comment]
Yeah, I'm happy to post about it on /r/reddit

**Comment by stefandollase** [comment]
Okay, I am mostly done with the release. You can find the new thread in the minecraftforum [here](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-tools/2626547-amidst-has-found-a-new-home). However, I did not yet create the second post in the seed subforum.

@BigAlanM You did not reply to my question in #33. You suggested that we should open a second thread in the seeds subforum. I agreed to this and suggested to create the post with the toolbox4minecraft account, since this account was created to be the owner of the Amidst posts: If I leave the project in the future, then I will pass the login credentials to another developer. This allows the project to keep the same thread instead of creating a new one. Furthermore, I asked you to suggest content for this second forum entry, however you did not react on this. Maybe you can clarify the situation?

Do you not like the idea to create the second post with the toolbox4minecraft account? I found your [current thread](http://www.minecraftforum.net/forums/minecraft-discussion/seeds/2361919-using-amidst-type-programs-to-map-and-research) in the seeds subforum. You seem to update the initial post quite often. Also, it refers to more tools than just Amidst. Is the new post supposed to replace your current one? This would explain why you don't like the idea of creating it with the toolbox4minecraft account. I am fine with you creating the second post in the seeds subforum, if it is about multiple tools, similar to the current one. But maybe I simply interpreted too much and you are cool with everything? Either way, please communicate with us and tell us your plans :-D

@Treer Thanks. Can you post the link here when you created the post?

**Comment by Treer** [comment]
I'm a tiny bit worried about [this change](https://github.com/toolbox4minecraft/amidst/commit/7dd9f228358aae67edeb1907fe5d66574f23e839#diff-70a542e04392ade1727ce4219e79abbcR62), I recall getting a null reference exception there because a widget managed to pass a null string. But I can't recall the details or reproduce it in Amidst 4. You refactored all the single line widgets to use updateTextLines() and so this might have fixed it at the source.

When I get home I'll see if it happens on a version that doesn't have your refactoring.

**Comment by stefandollase** [comment]
Yes, I noticed the null lines from the single line text widgets. This was one of the reasons to directly overwrite `updateTextLines()` in the concrete widgets instead of using the `updateText()` method. Since I did not find any other example where null is a line, I removed the check.

Also, since null is not supposed to be a line, it is not a good idea to defend against this case. This makes the code inconsistent, so at the end you don't even know whether null is a possible value. As soon as you see code which defends against this case you kind of have to think that null is indeed a valid line value. If you don't defend against cases which should not occur, the code will be easier to read and you are more likely to notice the wrong usage of the code during manual testing.

In short: I prefer the fail fast approach over the suppression of errors, because it helps to quickly find bugs during manual testing and it makes the code easier to follow.

Anyway, I would be happy if you did verify there is no sneaky bug.

**Comment by Treer** [comment]
Ah excellent. I also prefer fast fail.

**Comment by Treer** [comment]
I was going to wait before posting on reddit, but then I saw [this post](https://www.reddit.com/r/Minecraft/comments/48d5y2/spigot_bukkit_craftbukkit_bungeecord_for_19_now/), and that seemed like a good way to do it.

So, [here's an announcement on reddit](https://www.reddit.com/r/Minecraft/comments/48e3he/amidst_for_19_now_available_locate_igloos/)

**Comment by BigAlanM** [comment]
@stefandollase I apologize I didn't see (or forgot to respond to) your question in #33.

As long as provisions are made to pass on responsibility to maintain the first post of the thread and Wiki, I am ambivalent as to where it's located.  I only wanted it to be in the PC Seeds forum so that, as a moderator in the Discussion forum section, I could make any edits if necessary. 

As a matter of fact I do like the idea to create the second post with the toolbox4minecraft account.

Note: If you aren't familiar with the structure of the Minecraft forum, it's broken up into several sections, mostly by hardware platforms and are so named. EXCEPT for the first (and oldest) section which is just called Discussion.  That Discussion SECTION is further broken up into sub-forums including one named Discussion, one labeled Seeds, etc.
Because the Discussion Section is the first one listed when someone opens the Forums main index page new members are very likely to post to these sub forums that were the ORIGINAL ones when there was only a PC version of Minecraft.  As a moderator, I spend a good deal of time moving new threads to their appropriate hardware specific sub forums.  
My thread and post editing powers ONLY extend to the Discussion Section and, of course, my own posts and thread titles.
These "powers" will only exist as long as I'm a forum moderator.  Your guess is as good as mine as to how long that will be.

My thread in PC Seeds is really just an explanation for others on how I, personally, quickly evaluate posted seeds there and what I use to search for requested seeds.  As I get familiar with other utility programs that thread will probably change.

Once the new thread and wiki are up I will be adding links to both in that thread in seeds and in my forum signature.

**Comment by stefandollase** [comment]
Thanks for the verification @Treer.

Also, thanks for creating the reddit thread. I think the timing was good. There is also [another reddit thread](https://www.reddit.com/r/Minecraft/comments/47vg3v/does_amidst_work_with_19/), which routed quite some traffic to this repository. BTW: Can you see [the traffic statistics for this repository](https://github.com/toolbox4minecraft/amidst/graphs/traffic)? The most traffic currently comes from the old minecraftforum thread and the reddit thread linked above. The most popular content is the releases page, followed by the repository main page and the issues overview.

Regarding the link to the forum thread. I guess you cannot edit reddit posts? Maybe, you can add another post which refers to the forum thread. Maybe, you can also post the collection of links to the other reddit thread?

I am going to create a short second thread in the seeds subforum, soon. It will also contain links to at least the repository main page, the releases page and the primary thread.

Next, it might help the new primary minecraftforum thread if both of you react on it by replying something. This shows the community that it is indeed created by the Amidst developers and not just the new random avatar-less account ;-) Also, it shows that you support the moving of the project and that it is not just another small fork of the original Amidst, but indeed the successor of it. However, this is just my thought. You can do whatever you think is right ;-)

@BigAlanM Thanks for the clarification and explanation :-) So, I am going to create the second thread, independently of your current thread in the seed subforum.

**Comment by stefandollase** [comment]
I just created the second thread. You can find it [here](http://www.minecraftforum.net/forums/minecraft-discussion/seeds/2627219-amidst-now-supports-minecraft-1-9-and-has-a-new).

**Comment by stefandollase** [comment]
Just an update about the statistics: We got many visitors from reddit (2000), followed by the minecraftforum (600), and the minecraft wiki (200). The most popular content is still the release page (2000), but the main page is pretty close (2000). I made the numbers very rough.

I have to say, I am positively surprised that there are no new bug reports until now. I was a bit worried about that. I guess this is the right time to say thanks to all of you :-)

Nevertheless, I will now take a break from the active Amidst development. It took lots of my time for the last few month and I think it is now in a good state. Of course, there are lots of feature requests waiting for their implementation, but I currently cannot spend my time for this. I will be inactive at least until April. However, I will review new issues from time to time.

Even though the release is done, I will leave this issue open for a few more days.

**Comment by Treer** [comment]
I was surprised by how much work you put in to this. 

> I think it is now in a good state

Yeah, Amidst is back up to date.

I will be slowing down too.

**Comment by stefandollase** [comment]
Yes, I was surprised by myself. However, I really enjoy doing this. I learned a lot :-) But now I have other stuff to do.

I guess we have to wait a bit until @skiphs reacts and executes the tasks.

Since we might need to release bug-fix releases, we should probably move the complete integration of AmidstExporter to v5.0. This enables us to publish as much minor releases as we need to.

Nevertheless, I am closing this, since the release of v4.0 is done for us.
